drafts/2019-03-15-integration-testing-with-testcontainers.adoc

```diff
@@ -0,0 +1,317 @@
+=  Easy integration testing for JavaEE and MicroProfile applications with Testcontainers
+Andy Guibert <https://github.com/aguibert>
+
+Have you ever toiled with creating mock objects for unit tests? How about custom setup steps for integration tests? Ever had an issue in production because of differences in behavior between testing and production environments?
+
+If you answered yes to any of these questions, answer one more for me: do your applications run in docker containers?
+
+One of the great benefits of docker is that we get a nice consistent package that contains everything down to the OS, meaning it's portable to any hardware. Great, so lets leverage this to get consistent testing environments too!
+
+== Introducing Testcontainers
+
+image::https://d33wubrfki0l68.cloudfront.net/a661dbbe55be3e9cb77889f24835a44c6daf53c2/ce0aa/logo.png[Testcontainers logo, align="center"]
+
+https://www.testcontainers.org/[Testcontainers] is a great open source library that essentially wraps Docker in a Java API, and has first-class integration with JUnit. It is mainly used for external resource dependencies, and has first-class support for many popular databases such as Postgres and MySQL. In addition, it also has generic support for any docker container.
+
+Using the `GenericContainer` API, we can get a running Liberty server in no time!
+
+==== Gradle dependencies
+[source,groovy]
+----
+dependencies {
+  testCompile group: 'junit', name: 'junit', version: '4.12'
+  testCompile group: 'org.testcontainers', name: 'testcontainers', version: '1.10.6'
+}
+----
+
+==== JUnit test class
+[source,java]
+----
+import org.testcontainers.containers.GenericContainer;
+
+public class SampleTest {
+
+    @ClassRule
+    public static GenericContainer libertyContainer = new GenericContainer("open-liberty:microProfile2")
+                    .withExposedPorts(9080)
+                    .waitingFor(Wait.forHttp("/"));
+
+}
+----
+
+To break this down a bit:
+
+* JUnit's `@ClassRule` on any Testcontainer container object will result in the container being started before any `@Test` methods run, and then stopped after all tests are done running.
+* We need to declare what ports to expose from the container -- in this case 9080, as that is Liberty's default HTTP port
+* Using `waitFor()` we declare the criteria for the container being "ready". This check must pass before any tests start running. There are a few approaches to wait stratigies, which are https://www.testcontainers.org/features/startup_and_waits/[outlined in more detail in the Testcontainers documentation].
+
+== Testing my application's container
+
+Being able to start a Liberty container easily from a JUnit test is ok, but we need to get our test application in the container too. Since I am already putting my application in a Docker container, I have this Docker file on-hand already (it's the same one I use for production)
```

Comment:
Having this application already packaged for a docker container is key to this whole exercise, right?  So, if I'm new to this and I currently don't have this packaging in place, where can I look?  Will you have a github repo with this sample code so that I don't have invent something if I'm still in education mode?